### ISSUE TITLE
How to monitor container-name/ID on docker swarm architecture on multiple nodes ?

### ISSUE THREAD

**Issue by zouzou33** [louislam/uptime-kuma#3069]
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### 📝 Describe your problem

Hello,

Sorry in advance if I make any mistake or if my question is obvious to you. I am new on github, uptime kuma and pretty new on Docker Swarm as well. I am spent a lot of time to find out a solution but I'm blocked so here I am.

First of all here is my architecture :
Docker-swarm with 1 manager et 2 workers. Each node (even manager) run some containers. I use two overlay networks(1 frontend ,1 backend) and the ingress mesh routing by defaults. I also use glusterFs to share volume across different nodes.

```
+------------+   +-------------+   +-------------+
|   Manager  |   |   Worker 1  |   |   Worker 2  |
|+----------+|   |+-----------+|   |+-----------+|
||uptime-kuma||  || uptime-kuma||  || uptime-kuma||
|| Container1||  || Container3 ||  || Container5 ||
|| Container2||  || Container4 ||  || Container6 ||
|+---------+|    |+-----------+|   |+-----------+|
+-----------+    +-------------+   +-------------+
     \/ 		        \/ 			     \/ 
 +------------------------------------------------+
 | 				    Volumes     				  |
 | 				   (GlusterFS)  				  |
 | 				                				  |
 | 				/mnt/docker-data/ on another hdd  |
 +------------------------------------------------+
```

What I want to do ?
I would like to be able to monitor all my containers through their Name/ID whatever which node they are running. I am talking about the menu "Monitor Type" -> "Docker Container" -> "Container Name/ID". I have a lack of knowledge. I don't know if what I want to do is feasible.
Currently what I understand is : if uptime-kuma is deployed on the manager it will be able to monitor via the "container_name" only containers hosted on the manager because the "/var/run/docker.sock" is shared only with manager host. 

I have tried to use the mode: global in my stack but same behavior, i am only able to monitor container_name on the manager but not on Worker 1 & Worker2.

Here below, an exemple of my docker stack

herebelow my docker stack

version: '3.8'

services:
  uptime-kuma:
    image: louislam/uptime-kuma:1
    container_name: uptime-kuma
    volumes:
      - /mnt/docker-data/uptime-kuma-data:/app/data
      - /var/run/docker.sock:/var/run/docker.sock
    ports:
      - 3001:3001  
    restart: always
    deploy:
      mode: global
    networks:
      - test_frontend
      - test_backend

networks:
  test_frontend:
    external: true
  test_backend:
    external: true		

Thanks in advance for your lights.

### 🐻 Uptime-Kuma Version

1.21.2

### 💻 Operating System and Arch

Ubuntu 22.04.2 LTS 64bit

### 🌐 Browser

Firefox 111.0..1

### 🐋 Docker Version

 23.0.3

**Comment by zouzou33** [comment]
Hello,
I have tried to do an api call from the manager (cf attachement) but containers outside the manager is not detected. 
Best regards,
![api_docker](https://user-images.githubusercontent.com/2815828/232738278-061c5d48-ad3c-45e5-b34c-89f6ea55bc79.png)

**Comment by zouzou33** [comment]
Hello,
Allow me to update this post. So, no one has an idea to do what I explain in my initial post?

Is it even possible for a swarm architecture?
Thanks you.

**Comment by zouzou33** [comment]
Hello,
Allow me to update this post. So, no one has an idea to do what I explain in my initial post?

Is it even possible for a swarm architecture?
Thanks you.

**Comment by qnlbnsl** [comment]
Hi, is there any update on this?

**Comment by imoize** [comment]
I'm running docker swarm with 3 nodes and Ceph as storage distribution but I don't have any problem with database running on CephFS.

I had the same problem a few weeks ago, but only the monitoring part in the docker swarm cluster to monitor the service, in docker swarm I can't use the container name because sometimes the name changes.

To solve this problem I changed the `Monitor Type` to PING or TCP

And the `Hostname` to the Service name.

To find out the service name by running:
```bash
docker service ls
```

To use this method I have to use Adguard or Pihole DNS server, otherwise the ISP will try to resolve the Hostname and reply with a random ip. so the Monitor will detect that the container is still running, even though it is not.

So far it is still working fine in Swarm mode until now. And reports if the container is Down or Up correctly.

This may not be the right way to do this. Thanks

**Comment by djn84** [comment]
I could never get the docker integration working due to container ID/names changing constantly with a swarm deployment, but  discovered a decent workaround:

Using HTTP - Keyword monitors that query the swarm task states of the docker services...

**URL**
`http://<manager node ip>:2375/tasks?filters={"desired-state":["running"],"service":["service_name"]}`

**Keyword**
`"State":"running"`

That way if the running task fails on that service then the monitor will no longer be able to find one in a running state and the .

**Comment by scyto** [comment]
> I could never get the docker integration working due to container ID/names changing constantly with a swarm deployment, but discovered a decent workaround:
> 
> Using HTTP - Keyword monitors that query the swarm task states of the docker services...
> 
> **URL** `http://<manager node ip>:2375/tasks?filters={"desired-state":["running"],"service":["service_name"]}`
> 
> **Keyword** `"State":"running"`
> 
> That way if the running task fails on that service then the monitor will no longer be able to find one in a running state and the .

This works well for a service with only 1 replica

However when there are 3 replicas its breaks down as, say, one replica failing will mean the HTTP response has 2 instances of state:running and one of state:failed in the output.

Do you have a suggestion on how to mitigate?

**Comment by CommanderStorm** [comment]
If the response of said api is json, you could use the `json-query` monitor.
This might also be a good solution (have not checked it out) https://github.com/louislam/uptime-kuma/issues/957#issuecomment-1879761339

**Comment by SheepReaper** [comment]
Yeah, so i looked at this for a bit. No solutions, but posting some observations.

The containers API only lists containers running on that node. And the services API contains the container _template_ but no references to the replicated containers.

What you actually need is the tasks API (as mentioned above). (I was looking at this because I wanted to use autokuma to dynamically register the monitors via labels), the task information includes the container id, you could stop here and just use the Status.State, or you could call the container API on a container running on the host that's actually running the container in question. This might be simple if you're on a global deployed service with access to every node, but more likely you have a socket-proxy running, for security reasons. And you don't usually run the socket-proxy globally either. So I think here the preference is to use the tasks API since you can call any node for it and get the information you need.

The problem is that uptime-kuma needs the container id, which means you have to let kuma see socket-proxies on each node...

So in kuma's case, they need to implement a "swarm" monitor which uses the task API so that this can work without having to run a proxy on every node.

But this might be overkill, since we expect replicas to go down and rotate while the "service" is still virtually up, so in certain scenarios, you'd have to delete a task monitor whenever the task (swarm container) completes. So the multiple and rotating container id's are intentional, and are kind of the point of running a swarm/replicas, and conceptually doesn't align with what a _swarm_ considers "up".

So with that in mind, what we _really_ _really_ need is a swarm(or not swarm, services mean the same thing in non-swarm docker) _service_ monitor. *I propose that kuma considers "up" a service that has at least 1 replica in the "running" status* This can be achieved through the use of the service and tasks API (or make it a requirement to add a label to the container template and then you can skip using the services API). (OR, if you know the service ID, you can also skip the services API, but this is more brittle, since the service id changes if you delete a stack and re-deploy it)

Each service has a .Spec.Name, which you could configure in the monitor definition (Service name, OR ID, see above), call the services API with the service name and capture the service id (or skip this if you already have it). then call the tasks API and count the number of tasks that have that have a matching .ServiceID and the .Status.State is "running"

Tl;Dr: yes, there's some hoops you can jump through if you want to "monitor" containers on a swarm, without making code changes to kuma, but this is probably not what you want. If you want to use autokuma, be prepared to use a global socket proxy is you want to define docker monitors, but again, this is not conceptually correct.

Notes: autokuma allows you to pull service data with the templating stuff, but the docker stack compose parser (go) makes it impossible to define snippets, so just expect to skip snippets and define labels in both the service (container template) and deploy (service labels). Service labels are handy for defining HTTP monitors

@louislam @BigBoot

**Comment by SheepReaper** [comment]
> Yeah, so i looked at this for a bit. No solutions, but posting some observations.
> 
> The containers API only lists containers running on that node. And the services API contains the container _template_ but no references to the replicated containers.
> 
> What you actually need is the tasks API (as mentioned above). (I was looking at this because I wanted to use autokuma to dynamically register the monitors via labels), the task information includes the container id, you could stop here and just use the Status.State, or you could call the container API on a container running on the host that's actually running the container in question. This might be simple if you're on a global deployed service with access to every node, but more likely you have a socket-proxy running, for security reasons. And you don't usually run the socket-proxy globally either. So I think here the preference is to use the tasks API since you can call any node for it and get the information you need.
> 
> The problem is that uptime-kuma needs the container id, which means you have to let kuma see socket-proxies on each node...
> 
> So in kuma's case, they need to implement a "swarm" monitor which uses the task API so that this can work without having to run a proxy on every node.
> 
> But this might be overkill, since we expect replicas to go down and rotate while the "service" is still virtually up, so in certain scenarios, you'd have to delete a task monitor whenever the task (swarm container) completes. So the multiple and rotating container id's are intentional, and are kind of the point of running a swarm/replicas, and conceptually doesn't align with what a _swarm_ considers "up".
> 
> So with that in mind, what we _really_ _really_ need is a swarm(or not swarm, services mean the same thing in non-swarm docker) _service_ monitor. **I propose that kuma considers "up" a service that has at least 1 replica in the "running" status** This can be achieved through the use of the service and tasks API (or make it a requirement to add a label to the container template and then you can skip using the services API). (OR, if you know the service ID, you can also skip the services API, but this is more brittle, since the service id changes if you delete a stack and re-deploy it)
> 
> Each service has a .Spec.Name, which you could configure in the monitor definition (Service name, OR ID, see above), call the services API with the service name and capture the service id (or skip this if you already have it). then call the tasks API and count the number of tasks that have that have a matching .ServiceID and the .Status.State is "running"
> 
> Tl;Dr: yes, there's some hoops you can jump through if you want to "monitor" containers on a swarm, without making code changes to kuma, but this is probably not what you want. If you want to use autokuma, be prepared to use a global socket proxy is you want to define docker monitors, but again, this is not conceptually correct.
> 
> Notes: autokuma allows you to pull service data with the templating stuff, but the docker stack compose parser (go) makes it impossible to define snippets, so just expect to skip snippets and define labels in both the service (container template) and deploy (service labels). Service labels are handy for defining HTTP monitors
> 
> @louislam @BigBoot

Turns out the tasks API accepts a filter parameter that accepts either the service name or id, so no need to call the service API, if all you care about is that at least 1 replica is running. I made a PR that implements "simple availability" If it's accepted, I plan to also make the more advanced "service health monitor" which cares that if a service is desired to have a certain number of replicas, then that many need to be running in other to be "up" or healthy in this case. 0 would be "down", and any number in between is "degraded" which is PENDING in kuma terms. But naturally this is more complicated, especially since services have 2 distinct deployment modes: global and replicated. where replicated has a specific number, global does not, but depends on the number of nodes in the swarm. (specifically the number of nodes that are eligible (constraints) to host the container.) So without digging deeper, this seems straightforward for replicated services, but less so for global.

**Comment by BigBoot** [comment]
Sorry to Hijack this issue. @SheepReaper could you please elaborate on 

> Notes: autokuma allows you to pull service data with the templating stuff, but the docker stack compose parser (go) makes it impossible to define snippets, so just expect to skip snippets and define labels in both the service (container template) and deploy (service labels). Service labels are handy for defining HTTP monitors

If possible create a issue in the autokuma repo if this is something that needs to be fixed.
